aptos-move/aptos-aggregator/src/delayed_change.rs

```diff
@@ -120,6 +122,148 @@ impl<I: Copy + Clone> DelayedChange<I> {
         use DelayedChange::*;
         use DelayedFieldValue::*;
 
+        // There are only few valid cases for merging:
+        // - next_change being AggregatorDelta, and prev_change being Aggregator Create or Delta
+        // - next_change being SnapshotDelta, and prev_dependent_change being Aggregator Create or Delta
+        // everything else is invalid for various reasons
+        match (&prev_change, next_change) {
+            (None, v) => Ok(v.clone()),
+            (_, Create(_)) => Err(PanicOr::from(code_invariant_error(
+                "Trying to merge Create with an older change. Create should always be the first change.",
+            ))),
+
+            // Aggregators:
+            (Some(Create(Aggregator(prev_value))), Apply(AggregatorDelta { delta: next_delta })) => {
+                let new_data = next_delta.apply_to(*prev_value)?;
+                Ok(Create(Aggregator(new_data)))
+            },
+            (Some(Apply(AggregatorDelta { delta: prev_delta })), Apply(AggregatorDelta { delta: next_delta })) => {
+                let new_delta = DeltaWithMax::create_merged_delta(prev_delta, next_delta)?;
+                Ok(Apply(AggregatorDelta { delta: new_delta }))
+            },
+
+            // Snapshots:
+            (Some(Create(Snapshot(_) | Derived(_)) | Apply(SnapshotDelta {..} | SnapshotDerived { .. })), _) => Err(PanicOr::from(code_invariant_error(
+                "Snapshots are immutable, previous change cannot be any of the snapshots type",
+            ))),
+            (_, Apply(SnapshotDerived { .. })) => Err(PanicOr::from(code_invariant_error(
+                "Trying to merge SnapshotDerived with an older change. Snapshots are immutable, should only ever have one change.",
+            ))),
+            (Some(Create(Aggregator(prev_value))), Apply(SnapshotDelta { delta: next_delta, .. })) => {
+                let new_data = next_delta.apply_to(*prev_value)?;
+                Ok(Create(Snapshot(new_data)))
+            },
+            (Some(Apply(AggregatorDelta { delta: prev_delta })), Apply(SnapshotDelta { delta: next_delta, base_aggregator })) => {
+                let new_delta = DeltaWithMax::create_merged_delta(prev_delta, next_delta)?;
+                Ok(Apply(SnapshotDelta { delta: new_delta, base_aggregator: *base_aggregator }))
+            },
+        }
+    }
+}
+
+
+// TODO: See if we need these separate/duplicate classes or not
```

Comment:
I really really think we should not go down the path of having so many enums and duplication, it's very error-prone and hard to read / review / follow. Why not
```
struct DeltaOp {
  ...
  History,
}

enum History {
  Untracked, 
  Tracked(...)
}
```
and use deltas with untracked history? Or something similar.

Reply:
any such cleanup I will think about in the follow-up PR - I want to nail correctness here.

I haven't started with what you suggested, because we might want to remove DeltaOp instead, as History is going to be tracked separately - or at least remove the squashing need from DeltaOp - and with it remove the need for implementing DeltaHistory.merge 

similar reasoning with DelayedChange vs DelayedEntry